plugin/pulpcore/plugin/stages/artifact_stages.py

```diff
@@ -0,0 +1,274 @@
+    files and creates a new :class:`~pulpcore.plugin.models.Artifact` object from the downloaded
+    file and its digest data. The new :class:`~pulpcore.plugin.models.Artifact` is not saved but
+    added to the :class:`~pulpcore.plugin.stages.DeclarativeArtifact` object, replacing the likely
+    incomplete :class:`~pulpcore.plugin.models.Artifact`.
+
+    Each :class:`~pulpcore.plugin.stages.DeclarativeContent` is sent to `out_q` after all of its
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` objects have been handled.
+
+    This stage creates a ProgressBar named 'Downloading Artifacts' that counts the number of
+    downloads completed. Since it's a stream the total count isn't known until it's finished.
+
+    This stage drains all available items from `in_q` and starts as many downloaders as possible.
+
+    Args:
+        max_concurrent_downloads (int): The maximum number of concurrent downloads this stage will
+            run. Default is 100.
+
+    Returns:
+        An object containing the artifact_downloader stage to be included in a pipeline.
+    """
+
+    def __init__(self, max_concurrent_downloads=100):
+        self.max_concurrent_downloads = max_concurrent_downloads
+
+    async def stage(self, in_q, out_q):
+        """
+        Download undownloaded Artifacts, but don't save them in the db.
+
+        Args:
+            in_q (:class:`asyncio.Queue`): The queue to receive
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` objects from that may have
+                undownloaded files.
+            out_q (:class:`asyncio.Queue`): The queue to put
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` objects into, all of which have
+                files downloaded.
+
+        """
+        pending = set()
+        incoming_content = []
+        outstanding_downloads = 0
+        shutdown = False
+        saturated = False
+        with ProgressBar(message='Downloading Artifacts') as pb:
+            while True:
+                if not saturated:
+                    try:
+                        content = in_q.get_nowait()
+                    except asyncio.QueueEmpty:
+                        if not incoming_content and not shutdown and not pending:
+                            content = await in_q.get()
+                            incoming_content.append(content)
+                            continue
+                    else:
+                        incoming_content.append(content)
+                        continue
+
+                for i, content in enumerate(incoming_content):
+                    if content is None:
+                        shutdown = True
+                        continue
+                    downloaders_for_content = []
+                    for declarative_artifact in content.d_artifacts:
+                        if declarative_artifact.artifact._state.adding:
```

Comment:
Should not be accessing protected `_state` member.

Reply:
yup this isn't right. will fix in next push. Thanks!